### PR TITLE
Incremental clustering

### PR DESCRIPTION
**Pull request by ruiwanguk**
Here are a list of changes made to facilitate incremental clustering:
1. Add a new Hadoop job to covert spectra into clusters, also perform filtering and add an unique cluster id. 
2. Modify MajorPeak Hadoop job to take clusters as input 
3. Add a new Hadoop job that cluster by grouping clusters according to their cluster ids, so the clusters share the same id will be clustered. 
4. Refactored AbstractClusterReducer, the original version included many different logic, this has been broken into a several abstract classes. 
5. Change the order of the input arguments for all the jobs to enable multiple input directories
6. Add two new shell scripts for running incremental clustering jobs

### CODE REVIEW

File: src/main/java/uk/ac/ebi/pride/spectracluster/hadoop/merge/MergeClusterByIDReducer.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import uk.ac.ebi.pride.spectracluster.cluster.ICluster;
+import uk.ac.ebi.pride.spectracluster.hadoop.util.AbstractClusterReducer;
+import uk.ac.ebi.pride.spectracluster.hadoop.util.ConfigurableProperties;
+import uk.ac.ebi.pride.spectracluster.hadoop.util.IOUtilities;
+
+import java.io.IOException;
+import java.util.Collection;
+
+/**
+ * MergeClusterByIDReducer clusters using ClusteringEngine
```
Comment (jgriss) [inline src/main/java/uk/ac/ebi/pride/spectracluster/hadoop/merge/MergeClusterByIDReducer.java]: I'm not quite sure I understand the purpose of this job completely. It basically clusters all clusters that share the same id in one job? Why would clusters share the same id? Is this done just to make sure no duplicate ids are present? But I'm 99% certain it's me not getting something here!!!

File: src/main/java/uk/ac/ebi/pride/spectracluster/hadoop/util/AbstractIncrementalClusterReducer.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        super.cleanup(context);
+    }
+
+    protected abstract <T> void updateEngine(final Context context, final T key) throws IOException, InterruptedException;
+
+
+    /**
+     * Before writing out a cluster, remove all the non-fitting spectra and each out as a single-spectrum cluster
+     */
+    protected void writeCluster(Context context, ICluster cluster) throws IOException, InterruptedException {
```
Comment (jgriss) [inline src/main/java/uk/ac/ebi/pride/spectracluster/hadoop/util/AbstractIncrementalClusterReducer.java]: We could think about trying to cluster these non-fitting spectra again. Maybe they form a cluster. I'm not sure myself whether that's a sensible idea though.

**Comment by jgriss** [comment]
@ruiwanguk I think that's very nice work! Code looks very clean and structured! Congrats! I must admit though that I can't really assess all the Hadoop code properly. But here I completely trust you.

**Comment by ruiwanguk** [comment]
@jgriss When you have time, I can explain the code in detail to you. I hope it should be pretty straight forward to follow.

**Pull request merged**